content/blogposts/2023/slow-to-simd.md

```diff
@@ -0,0 +1,411 @@
+
+## The target
+
+This is a simple implementation of a function that calculates the dot product of two vectors. My goal is to outline the journey I took to optimize this function, and to share some tools I picked up along the way.
+
+```go
+func DotNaive(a, b []float32) float32 {
+	sum := float32(0)
+	for i := 0; i < len(a) && i < len(b); i++ {
+		sum += a[i] * b[i]
+	}
+	return sum
+}
+```
+
+Unless otherwise stated, all benchmarks are run on an Intel Xeon Platinum 8481C 2.70GHz CPU. This is a `c3-highcpu-44` GCE VM. The code in this blog post can all be found in runnable form [here](https://github.com/camdencheek/simd_blog).
+
+## Loop unrolling
+
+Modern CPUs do this thing called [_instruction pipelining_](https://en.wikipedia.org/wiki/Instruction_pipelining) where it can run multiple instructions simultaneously if it finds no data dependencies between them. A data dependency just means that the input of one instruction depends on the output of another.
+
+In our simple implementation, we have data dependencies between our loop iterations. A couple, in fact. Both `i` and `sum` have a read/write pair each iteration, meaning an iteration cannot start executing until the previous is finished.
+
+A common method of squeezing more out of our CPUs in situations like this is known as [_loop unrolling_](https://en.wikipedia.org/wiki/Loop_unrolling). The basic idea is to rewrite our loop so more of our relatively-high-latency multiply instructions can execute simultaneously. Additionally, it amortizes the fixed loop costs (increment and compare) across multiple operations.
+
+```go
+func DotUnroll4(a, b []float32) float32 {
+	sum := float32(0)
+	for i := 0; i < len(a); i += 4 {
+		s0 := a[i] * b[i]
+		s1 := a[i+1] * b[i+1]
+		s2 := a[i+2] * b[i+2]
+		s3 := a[i+3] * b[i+3]
+		sum += s0 + s1 + s2 + s3
+	}
+	return sum
+}
+```
+
+In our unrolled code, the dependencies between multiply instructions are removed, enabling the CPU to take more advantage of pipelining. This increases our throughput by 37% compared to our naive implementation.
+
+<Chart
+    rows={[
+        {name: "DotNaive", value: 941900},
+        {name: "DotUnroll4", value: 1286000},
+    ]}
+/>
+
+Note that we can actually improve this slightly more by twiddling with the number of iterations we unroll. On the benchmark machine, 8 seemed to be optimal, but on my laptop, 4 performs best. However, the improvement is quite platform dependent and fairly minimal, so for the rest of the post, I'll stick with an unroll depth of 4 for readability.
+
+## Bounds-checking elimination
+
+In order to keep out-of-bounds slice accesses from being [a security vulnerability](https://en.wikipedia.org/wiki/Heartbleed), the go compiler inserts checks before each read. You can check it out in the [generated assembly](https://go.godbolt.org/z/qT3M7nPGf) (look for `runtime.panic`).
```

Comment:
I don't believe it is correct to assert that out-of-bounds checking is purely a response to Heartbleed. As example, [Pascal featured bounds checking](https://en.wikipedia.org/wiki/Bounds_checking) and the language was developed in 1970 - before the Heartbleed discovery. 

However, if you'd like to you can use Hartbleed as motivation for why bounds checking is important.

Reply:
> use Hartbleed as motivation for why bounds checking is important

Oh, that was the intent. I'll reword that